src/transport_performance/utils/defence.py

```diff
@@ -169,6 +170,50 @@ def _bool_defence(some_bool, param_nm):
     return None
 
 
+def _string_defence(string, param_nm):
+    """Defence checking. Not exported."""
+    if not isinstance(string, str):
+        raise TypeError(f"'{param_nm}' expected str. Got {type(string)}")
+
+    return None
+
+
+def _integer_defence(some_int, param_nm):
+    """Defence checking. Not exported."""
+    if not isinstance(some_int, int):
+        raise TypeError(f"'{param_nm}' expected int. Got {type(some_int)}")
+
+    return None
+
+
+def _dict_defence(some_dict, param_nm):
+    """Defence checking. Not exported."""
+    if not isinstance(some_dict, dict):
+        raise TypeError(f"'{param_nm}' expected dict. Got {type(some_dict)}")
+
+    return None
+
+
+def _dataframe_defence(some_df, param_nm):
+    """Defence checking. Not exported."""
+    if not isinstance(some_df, pd.DataFrame):
+        raise TypeError(
+            f"'{param_nm}' expected pd.DataFrame. Got {type(some_df)}"
+        )
+
+    return None
+
+
+# main use case for this is to avoid
+# complexity limits in
+# GtfsInstance._plot_summary()
+def _string_and_nonetype_defence(some_value, param_nm):
+    if not isinstance(some_value, (str, type(None))):
+        raise TypeError(
+            f"'{param_nm}' expected type str. Found type {type(some_value)}"
+        )
```

Comment:
@CBROWN-ONS interested in this function. It's like `_string_defence` but additionally passes for None. I can see that `GtfsInstance._plot_summary` will play ball with x and y labels or without - nice. I can see 2 ways forward for this utility:
1. Update the TypeError message to mention nonetype.
2. Scrap this function, moving the logic into `_string_defence` with an additional parameter `raise_on_nonetype = True`.
Also worth considering whether any of the other defence utils would benefit from the same approach. Do you have a view on which would be better?

Reply:
Regarding point 1, I definitely think this should be the case.  
Moving onto point 2, I feel that we could pass a parameter to the defences to allow None types. This would also simplify the defence.py file as there wouldn't have to be multiple functions for each datatype defence.  Whether or no there are any other use cases throughout the package I'm not sure.

Reply:
Actually, can this replace all of the defence utilities introduced here? (Haven't pushed this up yet but I can do if you want to test it out).
```
def _type_defence(some_object, param_nm, types) -> None:
    """Internal defence checking utility. Can handle NoneType.

    Parameters
    ----------
    some_object : Any
        Object to test with isinstance.
    param_nm : str
        A name for the parameter. Useful when this utility is used in a wrapper
        to inherit the parent's parameter name and present in error message.
    types : type or tuple
        A type or a tuple of types to test `some_object` against.
    Raises
    ------
    TypeError
        `some_object` is not of type `types`. 

    """
    if not isinstance(some_object, types):
        raise TypeError(
            f"`{param_nm}` expected {types}. Got {type(some_object)}"
            )
    
    return None
```

I would suggest this one func could replace:
* _bool_defence
* _string_defence
* _integer_defence
* _dict_defence
* _dataframe_defence
* _string_and_nonetype_defence (by using it with a tuple of types, like `_type_defence("some string", "x label", (str, type(None))`)

Reply:
I thought of doing this before and I agree that it's a good idea.  
It also removes the need to add any additional type-checking defences in the future. I'm happy for you to push this assuming that it has been tested locally and retains the same behaviour as the previous defences.  

You'd have to update all instances of usage before pushing this, including functions impacted by `_bool_defence()` which was already present in `defences.py`.  

Overall, I'd be happy with this change, much better than several smaller functions!

Reply:
I just pushed a refactored version of this branch and raised a PR for your consideration @CBROWN-ONS :
https://github.com/datasciencecampus/transport-network-performance/pull/123

This branches off gtfs-html-new and wants to merge back to that branch.

Reply:
Tests pushed. Resolving this thread.